### ISSUE TITLE
Var declaration file *.auto.pkr.hcl does not load when specifying a template on build

### ISSUE THREAD

**Issue by MethylHaine** [hashicorp/packer#12138]
#### Overview of the Issue

This issue might come from my misunderstanding of the docs or the way variables are managed with Packer.

It seems files `*.auto.pkr.hcl` are not loaded when giving specific template as build target (e.g `dir/template.json.pkr.hcl`) :
`Error: Undefined -var variable`

It does load fine however when building with no specific target, such as `dir/.`.
As per the docs :
```
Packer also has the ability automatically load one or more variable definitions files if they are present:

    Any files with names ending in `.auto.pkrvars.hcl` or `.auto.pkrvars.json`.
```

#### Expected Behavior

Files with `*.auto.pkr.hcl` or `*.auto.pkrvars.hcl` should be loaded whether a specific target is given on build command or not. 

#### Reproduction Steps

I'm deploying templates on a VSphere instance with packer through a Gitlab CI.

Directory structure of my project is as follows :
```bash
.
├── .gitignore
├── .gitlab-ci.yml
├── Templates
│   ├── Linux
│   │   ├── template-RHEL7.json.pkr.hcl
│   │   ├── template-RHEL8.json.pkr.hcl
│   │   ├── variables.auto.pkr.hcl
│   └── Windows
└── vars
    ├── .gitlab-ci-jobs.yml
    └── .gitlab-ci-vars.yml
```

Running this build command is working fine and generates all templates in Linux dir :
```bash
packer build -force    \
-var "vsphere_user=${VSPHERE_DEV_USER}"         \
-var "vsphere_password=${VSPHERE_DEV_PASSWORD}" \
-var "vm_password=${TEMPLATE_ROOT_PASSWORD}"    \
Templates/Linux/.
```
This command however throws an undefined var error :
```bash
packer build -force    \
-var "vsphere_user=${VSPHERE_DEV_USER}"         \
-var "vsphere_password=${VSPHERE_DEV_PASSWORD}" \
-var "vm_password=${TEMPLATE_ROOT_PASSWORD}"    \
Templates/Linux/template-RHEL7.json.pkr.hcl
```
I have tried to add a `variables.auto.pkrvars.hcl` definition file, same behavior.
input vars declared in the `variables.auto.pkr.hcl` declaration file are not loaded when I try to build a specific template.

### Packer version

1.8.4

### Simplified Packer Template

```hcl
source "vsphere-iso" "template_rhel7" {
  CPUs                 = local.vm_cpu
  RAM                  = local.vm_ram_size
  RAM_reserve_all      = "false"
  boot_command         = ["XXXX"]
  boot_order           = "disk,cdrom"
  boot_wait            = "10s"
  cluster              = local.vsphere_cluster
  convert_to_template  = "true"
  datacenter           = local.vsphere_datacenter
  datastore            = local.vsphere_datastore
  disk_controller_type = ["pvscsi"]
  firmware             = "bios"
  folder               = "Templates"
  guest_os_type        = "rhel7_64Guest"
  insecure_connection  = "true"
  iso_paths            = [local.iso_path_rhel7]
  network_adapters {
    network      = local.vsphere_network
    network_card = "vmxnet3"
  }
  notes        = "VM template for RHEL 7.\nGenerated by Hashicorp™ Packer."
  password     = "${var.vsphere_password}"
  ssh_password = "${var.vm_password}"
  ssh_username = "xxxx"
  storage {
    disk_size             = local.vm_disk_size
    disk_thin_provisioned = "true"
  }
  username       = "${var.vsphere_user}"
  vcenter_server = local.vsphere_server
  vm_name        = local.vm_name_rhel7
}

build {
  sources = ["source.vsphere-iso.template_rhel7"]
}
```

### Operating system and Environment details

Running on `hashicorp/packer:latest` image in a dedicated runner as part of the Gitlab CI.

### Log Fragments and crash.log files

```
Error: Undefined -var variable
A "vsphere_user" variable was passed in the command line but was not found in
known variables. To declare variable "vsphere_user", place this block in one of
your .pkr files, such as variables.pkr.hcl
[...]
```

#### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by deric4** [comment]
I think you still need to define a _variable block_ defined somewhere based on this bit of documentation:


> Unlike legacy JSON templates the input variables within a variable definitions file must be declared via a variables block within a standard HCL2 template file *.pkr.hcl before it can be assigned a value. Failure to do so will result in an unknown variable error during Packer's runtime.
https://developer.hashicorp.com/packer/docs/templates/hcl_templates/variables#auto-loaded-variable-definitions-files

so something like
```hcl

variable "vsphere_user" {
  type = string
}

source "vsphere-iso" "template_rhel7" {
  /*
    snipped
  */
  username       = "${var.vsphere_user}"
  vcenter_server = local.vsphere_server
  vm_name        = local.vm_name_rhel7
}

build {
  sources = ["source.vsphere-iso.template_rhel7"]
}

```

**Comment by MethylHaine** [comment]
Hi @deric4 !

Thanks for your reply.
I should have said that I have indeed declared some _variables blocks_ in my `variables.auto.pkr.hcl` file :
```hcl
variable "vm_password" {
  type    = string
  default = "SENSITIVE"
  sensitive = true
}

variable "vsphere_password" {
  type    = string
  default = "SENSITIVE"
  sensitive = true
}

variable "vsphere_user" {
  type    = string
  default = "SENSITIVE"
  sensitive = true
}

locals {
  iso_path_rhel7        = "path_to_vsphere_iso"
  vm_cpu                = "2"
  vm_disk_size          = "40960"
  vm_ram_size           = "6144"
  vsphere_server        = "my-vsphere-server"
[...]
}
```
The variables declared in the _locals_ block are constants which need to be accessible from any of my templates.
The variables declared in the _variable_ blocks are inputs which are overridden with environment variables provided through the packer build command.

Now once again this config  is working fine when doing a packer build without a specific target,
but when specifying a target  the `variables.auto.pkr.hcl` file is not loaded :
```bash
packer build -force    \
-var "vsphere_user=${VSPHERE_DEV_USER}"         \
-var "vsphere_password=${VSPHERE_DEV_PASSWORD}" \
-var "vm_password=${TEMPLATE_ROOT_PASSWORD}"    \
Templates/Linux/template-RHEL7.json.pkr.hcl

Error: Undefined -var variable
A "vsphere_user" variable was passed in the command line but was not found in
known variables. To declare variable "vsphere_user", place this block in one of
your .pkr files, such as variables.pkr.hcl
[...]
```

It seems that when you specify a target packer only have knowledge about that target, and files like `*.auto.pkr.hcl`  are not auto-loaded anymore..

A workaround should be the `-var-file` option  :
> Given a file named `variables.pkrvars.hcl` Packer can be instructed to use the variable definitions file by using the `-var-file` command line flag.

But this also won't work because I can only pass a `*.pkrvars.hcl` file and such a definition file cannot be used to declare input variables..

```bash
packer build -force    \
-var "vsphere_user=${VSPHERE_DEV_USER}"         \
-var "vsphere_password=${VSPHERE_DEV_PASSWORD}" \
-var "vm_password=${TEMPLATE_ROOT_PASSWORD}"    \
-var-file="Templates/Linux/variables.auto.pkrvars.hcl"     \
Templates/Linux/template-RHEL7.json.pkr.hcl

Warning: Undefined variable
A "vsphere_datacenter" variable was set but was not declared as an input
variable. To declare variable "vsphere_datacenter", place this block in one of
your .pkr.hcl files, such as variables.pkr.hcl
[...]
```

In short what bothers me is that there should be the same behavior with variables whether I specify a target or not during build.

**Comment by MethylHaine** [comment]
It turns out that it is possible to achieve my goal with the following keywords :  
-  `only`
-  `except`

https://developer.hashicorp.com/packer/docs/templates/hcl_templates/onlyexcept

My issue was indeed coming from my lack of knowledge, as the following works :

_**template-RHEL7.json.pkr.hcl**_
```hcl
source "vsphere-iso" "template_rhel7" {
...
}

build {
  name = "my_build"
  sources = ["source.vsphere-iso.template_rhel7"]
}
```

**_gitlab-ci.yml_**
```bash
packer build -force  -only=my_build*            \
-var "vsphere_user=${VSPHERE_DEV_USER}"         \
-var "vsphere_password=${VSPHERE_DEV_PASSWORD}" \
-var "vm_password=${TEMPLATE_ROOT_PASSWORD}"    \
Templates/Linux/.
```

The variable declaration file `variables.auto.pkr.hcl` is correctly loaded and only the specified build is launched.
